File: cpp/encontrar_maximo.cpp
#include <algorithm>  // Permite usar max_element para obtener el valor de referencia.
#include <iostream>   // Permite mostrar informacion en la consola.
#include <stdexcept>  // Permite lanzar invalid_argument ante un arreglo vacio.
#include <vector>     // Permite almacenar los numeros en un vector dinamico.

int encontrarMaximo(const std::vector<int>& arr) {  // Recibe el vector por referencia constante y devuelve su maximo.
    if (arr.empty()) {  // Comprueba si el vector incumple la precondicion de no estar vacio.
        throw std::invalid_argument("El arreglo no puede estar vacio");  // Detiene la funcion con un mensaje explicativo.
    }  // Finaliza la validacion de la entrada.
    int maximo = arr[0];  // Toma el primer elemento como maximo provisional.
    for (std::size_t i = 1; i < arr.size(); ++i) {  // Recorre el vector desde el segundo elemento hasta el ultimo.
        if (arr[i] > maximo) {  // Compara el elemento actual con el mayor encontrado hasta ese momento.
            maximo = arr[i];  // Actualiza el maximo cuando el elemento actual es mayor.
        }  // Finaliza la condicion de actualizacion.
    }  // Finaliza el recorrido lineal del vector.
    return maximo;  // Devuelve el mayor valor despues de revisar todos los elementos.
}  // Finaliza la funcion encontrarMaximo.

int main() {  // Inicia el programa de prueba.
    const std::vector<int> datos = {34, -7, 12, 56, 23, 91, 4, 18, 67, 45, 0, 39, 72, 11, 8, 63, 27, 50, 16, 5, 88, 31, -12, 70};  // Crea el vector de prueba con 24 enteros.
    const int resultado = encontrarMaximo(datos);  // Ejecuta el algoritmo desarrollado.
    const int maximoReal = *std::max_element(datos.begin(), datos.end());  // Calcula la referencia con la biblioteca estandar.
    const bool coinciden = resultado == maximoReal;  // Verifica que los dos valores sean iguales.
    std::cout << "Arreglo (" << datos.size() << " elementos): [";  // Inicia la impresion del vector y de su tamano.
    for (std::size_t i = 0; i < datos.size(); ++i) {  // Recorre los elementos exclusivamente para mostrarlos.
        std::cout << datos[i];  // Imprime el elemento ubicado en la posicion actual.
        if (i + 1 < datos.size()) {  // Comprueba si todavia quedan elementos por imprimir.
            std::cout << ", ";  // Separa los elementos mediante una coma y un espacio.
        }  // Finaliza la condicion usada para separar los elementos.
    }  // Finaliza el recorrido de impresion.
    std::cout << "]\n";  // Cierra la representacion del vector y cambia de linea.
    std::cout << "Maximo calculado: " << resultado << '\n';  // Muestra el valor producido por encontrarMaximo.
    std::cout << "Maximo de referencia: " << maximoReal << '\n';  // Muestra el valor obtenido con max_element.
    std::cout << "Verificacion: " << (coinciden ? "CORRECTA" : "INCORRECTA") << '\n';  // Comunica el resultado de la comprobacion.
    return coinciden ? 0 : 1;  // Termina con exito solo cuando ambos resultados coinciden.
}  // Finaliza la funcion principal.

File: python/encontrar_maximo.py
"""Implementacion comentada del algoritmo encontrar_maximo para SIS210."""  # Describe el proposito del archivo.


def encontrar_maximo(arr):  # Define una funcion que recibe una secuencia no vacia.
    """Devuelve el mayor elemento de arr mediante un recorrido lineal."""  # Documenta el resultado de la funcion.
    if len(arr) == 0:  # Comprueba si la secuencia incumple la precondicion de no estar vacia.
        raise ValueError("El arreglo no puede estar vacio")  # Informa claramente el error si no hay elementos.
    maximo = arr[0]  # Toma el primer elemento como maximo provisional.
    for i in range(1, len(arr)):  # Recorre los indices desde el segundo elemento hasta el ultimo.
        if arr[i] > maximo:  # Compara el elemento actual con el mayor encontrado hasta ahora.
            maximo = arr[i]  # Actualiza el maximo solamente cuando aparece un valor mayor.
    return maximo  # Devuelve el mayor valor una vez revisado todo el arreglo.


datos = [34, -7, 12, 56, 23, 91, 4, 18, 67, 45, 0, 39, 72, 11, 8, 63, 27, 50, 16, 5, 88, 31, -12, 70]  # Crea el arreglo de prueba con 24 enteros.
resultado = encontrar_maximo(datos)  # Ejecuta el algoritmo desarrollado y guarda su respuesta.
maximo_real = max(datos)  # Obtiene un valor de referencia con la funcion estandar de Python.
coinciden = resultado == maximo_real  # Verifica si el resultado calculado coincide con la referencia.
print(f"Arreglo ({len(datos)} elementos): {datos}")  # Muestra los datos utilizados en la prueba.
print(f"Maximo calculado: {resultado}")  # Muestra el valor encontrado por el algoritmo.
print(f"Maximo de referencia: {maximo_real}")  # Muestra el valor obtenido mediante la biblioteca estandar.
print(f"Verificacion: {'CORRECTA' if coinciden else 'INCORRECTA'}")  # Comunica si ambas respuestas son iguales.

